clients/hadoopfs/src/main/java/io/lakefs/LinkOnCloseOutputStream.java

```diff
@@ -63,11 +64,14 @@ public void close() throws IOException {
 
         // Now the object is on the underlying store, find its parameters (sadly lost by
         // the underlying Hadoop FileSystem) so we can link it on lakeFS.
-        try {
-            lfs.linkPhysicalAddress(objectLoc, stagingLoc, physicalUri, metadataClient);
-        } catch (io.lakefs.clients.api.ApiException e) {
-            throw new IOException("link lakeFS path to physical address", e);
+        if (!this.isLinked) {
+            try {
+                lfs.linkPhysicalAddress(objectLoc, stagingLoc, physicalUri, metadataClient);
+            } catch (io.lakefs.clients.api.ApiException e) {
+                throw new IOException("link lakeFS path to physical address", e);
+            }
+            this.isLinked = true;
```

Comment:
This is a race, it may fail to trigger in some cases -- for instance when 2 threads both close the same LinkOnCloseOutputStreamOutputStream.  This will occur in some use-cases, and if the API allows multiple calls to close then it makes sense to require it to work.

If we believe that this fixes an actual bug then the fix is bad.  Otherwise it just adds a nicer error report and somewhat reduces the frequency of the bug... but does not remove it.